src/evaluators/JSXElement.js

```diff
@@ -251,13 +247,16 @@ function evaluateJSXAttributes(
             }
           }
         } else {
-          keyValue = undefined;
-          refValue = undefined;
-          containsAbstractSpreadAttribute = true;
+          abstractSpreadCount++;
           invariant(spreadValue instanceof AbstractValue || spreadValue instanceof ObjectValue);
 
-          if (!objectHasNoPartialKeyAndRef(realm, spreadValue)) {
-            mayContainRefOrKey = true;
+          if (
+            (spreadValue instanceof AbstractObjectValue &&
+              spreadValue.isSimpleObject() &&
+              spreadValue.isFinalObject()) ||
```

Comment:
Why does being simple and final imply it’s safe from having a key or a ref?

Reply:
The only place this can ever get set is the React reconciler with the props going into a function.

Reply:
What does “with the props going into a function” mean? Can you give an example?

Reply:
When the reconciler enters a component and it's completely abstract it creates completely abstract props. It makes them simple and final too.

Reply:
So how do we know they don't have ref/key?

Reply:
Let's use a more specific tagging mechanism for "doesn't have key/ref" because it seems orthogonal to finality.